### PR TITLE
FLUME-199 Remove hardcoded ports from unit tests

### PR DESCRIPTION
**Pull request by andrasbeni**
Many unit tests use hardcoded port numbers. This causes problems when running builds in parallel.

### CODE REVIEW

File: flume-ng-core/src/test/java/org/apache/flume/sink/TestThriftSink.java [hunk 1]
```diff
@@ -40,6 +40,9 @@
 
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.TrustManagerFactory;
+
+import java.io.IOException;
+import java.net.ServerSocket;
 import java.nio.charset.Charset;
 import java.util.Random;
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/sink/TestThriftSink.java]: This import for Random is not used.

File: flume-ng-core/src/test/java/org/apache/flume/sink/TestThriftSink.java [hunk 2]
```diff
@@ -40,6 +40,9 @@
 
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.TrustManagerFactory;
+
+import java.io.IOException;
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/sink/TestThriftSink.java]: Imported IOException is not used

File: flume-ng-sinks/flume-ng-kafka-sink/src/test/java/org/apache/flume/sink/kafka/util/TestUtil.java [hunk 3]
```diff
@@ -22,8 +22,10 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
 import java.net.BindException;
 import java.net.InetAddress;
+import java.net.ServerSocket;
 import java.net.UnknownHostException;
 import java.util.List;
 import java.util.Properties;
```
Comment (mcsanady) [inline flume-ng-sinks/flume-ng-kafka-sink/src/test/java/org/apache/flume/sink/kafka/util/TestUtil.java:31]: The import for Random in the next line can be deleted

File: flume-ng-core/src/test/java/org/apache/flume/instrumentation/http/TestHTTPMetricsServer.java [hunk 4]
```diff
@@ -37,19 +37,23 @@
...
- *
- */
 public class TestHTTPMetricsServer {
 
   Channel memChannel = new MemoryChannel();
   Channel pmemChannel = new PseudoTxnMemoryChannel();
   Type mapType = new TypeToken<Map<String, Map<String, String>>>() {}.getType();
   Gson gson = new Gson();
 
+  private static final int getFreePort() throws Exception {
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/instrumentation/http/TestHTTPMetricsServer.java]: "final" is not needed at private methods

File: flume-ng-core/src/test/java/org/apache/flume/source/TestAvroSource.java [hunk 5]
```diff
@@ -254,6 +233,12 @@ private void doRequest(boolean serverEnableCompression, boolean clientEnableComp
         source.getLifecycleState());
   }
 
+  private static int getFreePort() throws IOException {
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestAvroSource.java:293]: "final" is not needed at private methods

Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestAvroSource.java:293]: This method should not be redeclared in different packages.
If you can extract it outside the original package the private constraint cannot be used. There the "final" label will be useful.

File: flume-ng-core/src/test/java/org/apache/flume/source/TestNetcatSource.java [hunk 6]
```diff
@@ -52,6 +52,14 @@
   private static final Logger logger =
       LoggerFactory.getLogger(TestAvroSource.class);
 
+
+  private static final int getFreePort() {
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestNetcatSource.java]: Extract this from modules.
"final" is not necessary on private methods

File: flume-ng-embedded-agent/src/test/java/org/apache/flume/agent/embedded/TestEmbeddedAgent.java [hunk 7]
```diff
@@ -225,17 +225,8 @@ public Status appendBatch(List<AvroFlumeEvent> events)
   }
 
   private static int findFreePort() throws IOException {
-    ServerSocket socket = null;
-    try {
-      socket = new ServerSocket(0);
+    try (ServerSocket socket = new ServerSocket(0);) {
```
Comment (mcsanady) [inline flume-ng-embedded-agent/src/test/java/org/apache/flume/agent/embedded/TestEmbeddedAgent.java]: Unnecessary semicolon

File: flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/KafkaSourceEmbeddedKafka.java [hunk 8]
```diff
@@ -41,8 +42,16 @@
...
-  int serverPort = 18922;
+  private static int findFreePort() {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      return socket.getLocalPort();
+    } catch (IOException e) {
+      throw new AssertionError("Can not find free port.", e);
+    }
+  }
+
+  int zkPort = findFreePort(); // none-standard
```
Comment (mcsanady) [inline flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/KafkaSourceEmbeddedKafka.java]: zkPort and serverPort can be private

File: flume-ng-tests/src/test/java/org/apache/flume/test/util/StagedInstall.java [hunk 9]
```diff
@@ -78,6 +79,12 @@ public static synchronized StagedInstall getInstance() throws Exception {
     return INSTANCE;
   }
 
+  public static int findFreePort() throws IOException {
```
Comment (mcsanady) [inline flume-ng-tests/src/test/java/org/apache/flume/test/util/StagedInstall.java:82]: Can be extracted to another package

File: flume-ng-clients/flume-ng-log4jappender/src/test/java/org/apache/flume/clients/log4jappender/TestLog4jAppender.java [hunk 10]
```diff
@@ -54,9 +55,15 @@
   private Channel ch;
   private Properties props;
 
+  private static int getFreePort() throws Exception {
```
Comment (mcsanady) [inline flume-ng-clients/flume-ng-log4jappender/src/test/java/org/apache/flume/clients/log4jappender/TestLog4jAppender.java:63]: Can be extracted to another package

File: flume-ng-clients/flume-ng-log4jappender/src/test/java/org/apache/flume/clients/log4jappender/TestLog4jAppenderWithAvro.java [hunk 11]
```diff
@@ -58,14 +59,21 @@
   private AvroSource source;
   private Channel ch;
   private Properties props;
+  private int port;
+
+  private static int getFreePort() throws Exception {
```
Comment (mcsanady) [inline flume-ng-clients/flume-ng-log4jappender/src/test/java/org/apache/flume/clients/log4jappender/TestLog4jAppenderWithAvro.java:65]: Extract it

File: flume-ng-core/src/test/java/org/apache/flume/source/TestMultiportSyslogTCPSource.java [hunk 12]
```diff
@@ -56,17 +57,17 @@
...
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import static org.mockito.Mockito.*;
 
 public class TestMultiportSyslogTCPSource {
 
-  private static final Logger logger =
-      LoggerFactory.getLogger(TestMultiportSyslogTCPSource.class);
+  private static final int getFreePort() throws IOException {
```
Comment (mcsanady) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestMultiportSyslogTCPSource.java:65]: extract it

File: flume-ng-node/src/test/java/org/apache/flume/source/TestNetcatSource.java [hunk 13]
```diff
@@ -70,11 +70,19 @@ public TestNetcatSource(boolean ackForEveryEvent) {
   }
 
   @Parameters
-  public static Collection data() {
+  public static Collection<?> data() {
     Object[][] data = new Object[][] { { true }, { false } };
     return Arrays.asList(data);
   }
 
+  private static int getFreePort() {
```
Comment (mcsanady) [inline flume-ng-node/src/test/java/org/apache/flume/source/TestNetcatSource.java:78]: extract it

**Review by mcsanady — COMMENTED** [review]
Thank you for the patch.
I made a few suggestions.

**Comment by andrasbeni** [comment]
@mcsanady I removed unnecessary imports and keywords. But I could not find a place to move the getFreePort methods. As far as I can see, there is no module in Flume that all test modules depend on. Should I create a test-utils module?

**Review by mcsanady — APPROVED** [review]

**Comment by andrasbeni** [comment]
@adenes Given Miklos approved the PR, can you please commit?
Thanks

**Review by szaboferee — APPROVED** [review]
Tests were ok,
We will find a common place for a tool later.

**Comment by rgoers** [comment]
org.apache.flume.source.TestMultiportSyslogTCPSource failed for me withjava.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at java.net.Socket.<init>(Socket.java:434)
	at java.net.Socket.<init>(Socket.java:244)
	at org.apache.flume.source.TestMultiportSyslogTCPSource.testMultiplePorts(TestMultiportSyslogTCPSource.java:130)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

**Pull request closed without merging**